### PR TITLE
Use SSA for DisruptionTarget Pod conditions

### LINKED ISSUE #121678: Use SSA for DisruptionTarget Pod conditions
### What would you like to be added?

Use SSA instead of Patch.

Essentially, revert https://github.com/kubernetes/kubernetes/pull/121103
Although it might be desirable to leave Update for the disruption target garbage collector.

/sig apps

### Why is this needed?

With #113482 fixed, we can use SSA again.

### PR DESCRIPTION
**Pull request by kaisoz**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #121678 

#### Does this PR introduce a user-facing change?

```release-note
Use Server Side Apply (SSA) for managing `DisruptionTarget` conditions on Pods.
```

### CODE REVIEW

**Comment by mimowo** [comment]
I would suggest using SSA in disruption.go for consistency. In this controller we currently use `Update` just so that it fails in case there was a concurrent update (based on ResoureVersion bump). This semantics can be expressed using SSA and passing the ResourceVersion as in the reverted PR.

**Comment by kaisoz** [comment]
> I would suggest using SSA in disruption.go for consistency. In this controller we currently use `Update` just so that it fails in case there was a concurrent update (based on ResoureVersion bump). This semantics can be expressed using SSA and passing the ResourceVersion as in the reverted PR.

Good point! Reverted, thanks @mimowo

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Reverted to using Server Side Apply (SSA) instead of Patch for adding PodDisruptionConditions.
+Reverted to using Server Side Apply (SSA) instead of JSON Patch for managing `DisruptionTarget` conditions on Pods.
```

Was it JSON patch? I'm not sure what the old format was. The actual verb is still PATCH at the HTTP level, right?

**Comment by mimowo** [comment]
> as it JSON patch? I'm not sure what the old format was.

We used JSON Patch, but also Update for one controller: https://github.com/kubernetes/kubernetes/blob/7897910469aa091ebf6576740d055a7137fa147c/pkg/controller/disruption/disruption.go#L777. 
At the HTTP level I believe SSA and JSON Patch are PATCH requests.

Maybe the change log can be just:
```
Use Server Side Apply (SSA) for managing `DisruptionTarget` conditions on Pods.
```

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			},
 		}
 		if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
-			role.Rules = append(role.Rules, rbacv1helpers.NewRule("patch", "update").Groups(legacyGroup).Resources("pods/status").RuleOrDie())
```
Comment (mimowo) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:143]: I don't think we want to take the permission back, once given, there could be other things in the meanwhile that started to rely on it.

Comment (kaisoz) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go]: Make sense. Reverted it 👍🏻

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml [hunk 2]
```diff
@@ -436,7 +436,6 @@ items:
     - pods/status
     verbs:
     - patch
-    - update
```
Comment (mimowo) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:439]: Please also revert this line

Comment (kaisoz) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: Sorry I don't know how I missed that 🤦🏻

File: pkg/controller/disruption/disruption.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
+	corev1apply "k8s.io/client-go/applyconfigurations/core/v1"
```
Comment (mimowo) [inline pkg/controller/disruption/disruption.go:37]: The PR currently does not revert the change for Taint manager, but I think it should. The code is now moved to https://github.com/kubernetes/kubernetes/blob/c9158e9a19ef8aaf6c53f6293d1f86371fe8e8a6/pkg/controller/tainteviction/taint_eviction.go#L139

Comment (mimowo) [reply pkg/controller/disruption/disruption.go:37]: Oh, I see we have the code in taint_eviction. Not sure how I missed it or it was added later. Feel free to resolve.

**Review by mimowo — COMMENTED** [review]
/lgtm

**Comment by mimowo** [comment]
cc @liggitt 
would you like to review?

**Comment by mimowo** [comment]
cc @alculquicondor

File: pkg/controller/disruption/disruption.go (hunk 3 above)
Comment (kaisoz) [reply pkg/controller/disruption/disruption.go:37]: Thanks for the review @mimowo , tbh I almost missed this one 😓

**Comment by kaisoz** [comment]
@mimowo @alculquicondor does this still make sense? I'm asking because @dims labelled it as rotten

**Comment by dims** [comment]
rebase should have cleaned up the `rotten`... removed now

**Comment by kaisoz** [comment]
/remove-lifecycle stale
@mimowo @alculquicondor do you think this PR is still needed?

**Comment by mimowo** [comment]
> @mimowo @alculquicondor do you think this PR is still needed?

As in the [comment](https://github.com/kubernetes/kubernetes/issues/121678#issuecomment-2196337664).

**Comment by alculquicondor** [comment]
FYI https://github.com/kubernetes/kubernetes/pull/126644

I suppose this PR would fix that too.
